models/badges.js

```diff
@@ -27,36 +27,106 @@ const fetchBadges = async ({ size = 100, page = 0 }) => {
 
 /**
  * Fetches the data about user badges
- * @param query { Object }: Filter for badges data
- * @return {Promise<userBadgeModel|Array>}
+ * @param username { string }: Filter for badges data
+ * @return {Promise<{badges: Array}>}
  */
-
-const fetchUserBadges = async (username) => {
+async function fetchUserBadges(username) {
   try {
-    const userBadges = [];
-    let userExists = false;
+    let badges = [];
     const result = await fetchUser({ username });
-    if (result.userExists) {
-      userExists = true;
-      const userID = result.user.id;
-      const snapshot = await badgeModel.get();
-
-      snapshot.forEach((item) => {
-        if (item.data()?.users?.includes(userID)) {
-          const { title, description } = item.data();
-          userBadges.push({ title, description });
-        }
-      });
+    if (!result.userExists) {
+      return { userExists: false, badges };
     }
-
-    return { userExists, userBadges };
+    const userId = result.user.id;
+    const badgeIdsSnapshot = await userBadgeModel.where("userId", "==", userId).get();
+    const badgeDocReferences = badgeIdsSnapshot.docs.map((doc) => {
+      const badgeId = doc.get("badgeId");
+      return firestore.doc(`badges/${badgeId}`);
+    });
+    // INFO: getAll accepts unpacked array
+    const badgesSnapshot = await firestore.getAll(...badgeDocReferences);
+    badges = badgesSnapshot.map((doc) => convertFirebaseDocumentToBadgeDocument(doc.id, doc.data()));
+    return { userExists: true, badges };
   } catch (err) {
     logger.error("Error retrieving user badges", err);
     return err;
   }
-};
+}
+
+/**
+ * Add badge to firestore
+ * @param  badgeInfo { Object }: has badge name, description, imageUrl and createdBy
+ * @return {Promise<{id: string, createdAt: {date: string, time: string}}|Object>}
+ */
+async function createBadge(badgeInfo) {
+  try {
+    const createdAt = admin.firestore.Timestamp.now();
+    // INFO: check if description is missing
+    const description = badgeInfo.description ?? "";
+    const docRef = await badgeModel.add({
+      ...badgeInfo,
+      description,
+      createdAt,
+    });
+    const { date, time } = convertFirebaseTimestampToDateTime(createdAt);
+    const snapshot = await docRef.get();
+    const data = snapshot.data();
+    return { id: docRef.id, ...data, createdAt: { date, time } };
+  } catch (err) {
+    logger.error("Error creating badge", err);
+    return err;
+  }
+}
+
+/**
+ * assign badges to user
+ * @param { Object }: userId: string and badgeIds: Array<string>
+ * @return {Promise<{docIds: Array<string>}|Object>}
+ */
+async function assignBadges({ userId, badgeIds }) {
+  try {
+    const docIds = [];
+    const batch = firestore.batch();
+    badgeIds.forEach((badgeId) => {
+      const ref = userBadgeModel.doc();
+      const id = ref.id;
+      batch.create(ref, { userId, badgeId });
+      docIds.push(id);
+    });
+    await batch.commit();
+    return { docIds };
```

Comment:
The batch operation has a limit of 500 documents. we should add it in logger as well if we find more than 500 docs are there to update.

Reply:
Reffer below comment: https://github.com/Real-Dev-Squad/website-backend/pull/832#discussion_r1070247617